sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1050 @@
+trademarks of The Khronos Group Inc. OpenCL(TM) is a trademark of Apple Inc.
+used by permission by Khronos.
+
+
+This extension is written against the SYCL 2020 revision 5 specification. All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+== Contributors
+
+Pablo Reble, Intel +
+Julian Miller, Intel +
+John Pennycook, Intel +
+Guo Yejun, Intel +
+Ewan Crawford, Codeplay +
+Ben Tracy, Codeplay +
+Duncan McBain, Codeplay +
+Peter Žužek, Codeplay +
+Ruyman Reyes, Codeplay +
+Gordon Brown, Codeplay +
+Erik Tomusk, Codeplay +
+Bjoern Knafla, Codeplay +
+Lukas Sommer, Codeplay +
+Ronan Keryell, AMD +
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+== Introduction
+
+Through the use of command groups SYCL is already able to create a dependency
+graph (in the form of a directed acyclic graph) of kernel execution at runtime,
+as a command group object defines a set of requisites (edges) which must be
+satisfied for kernels (nodes) to be executed. However, because command-group
+submission is tied to execution on the queue, without having a prior
+construction step before starting execution, optimization opportunities are
+missed from the runtime not knowing the complete dependency graph ahead of
+execution.
+
+The following benefits would become possible if the user could define a
+dependency graph to the SYCL runtime prior to execution:
+
+* Reduction in runtime overhead by only submitting a single graph object, rather
+  than many individual commands.
+
+* Enable more work to be done offline, in particular producing a graph ahead of
+  time allows for improved performance at runtime from reduced overhead.
+
+* Unlock DMA hardware features through graph analysis by the runtime.
+
+* Whole graph optimizations become available, including but not limited to:
+** Kernel fusion/fission.
+** Inter-node memory reuse from data staying resident on device.
+** Identification of the peak intermediate output memory requirement, used for
+   more optimal memory allocation.
+
+As well as benefits to the SYCL runtime, there are also advantages to the user
+developing SYCL applications, as repetitive workloads no longer have to
+redundantly issue the same sequence of commands. Instead, a graph is only
+constructed once and submitted for execution as many times as is necessary, only
+changing the data in input buffers or USM allocations. For applications from
+specific domains, such as machine learning, where the same command group pattern
+is run repeatedly for different inputs, this is particularly useful.
```

Comment:
The problem with a DAG is the "A".
Amortisation of costs requires re-use, but invocation of the graph does not express whether further invocation will happen. The runtime can only do "loop unrolling" optimisations when it knows ahead of time that the graph (the loop body) will be executed multiple times (the loop count). For iterative algorithms, the output from one invocation of the graph will feed into the input of the next invocation of the graph (either directly or indirectly via some non-graph work).

Reply:
Could you clarify what you are proposing? Is it to remove the acyclic restriction on a `command_graph`, which would modify the SYCL execution model rather than expose it. Or some mechanism to submit a `command_graph` N times in a single API invocation, rather than N individual calls?

Reply:
tl;dr; both are interesting avenues to explore.

> some mechanism to submit a command_graph N times in a single API invocation, rather than N individual calls?

This mechanism already exists? Pseudo-code:
```
Create Q_R; put it into "recording" mode targeting graph G_N
Do N times (submit graph G_1 to Q_R)
Finalise graph G_N
Create Q; // default "executable" mode
Do 1 times (submit graph G_N to Q)
```

Even without SYCL graph API, the user could (in some circumstances) manually fuse N kernel invocations.

If the user can submit N kernels concurrently that each do 1 kernel body task:
``` C++
queue q;
void* kernel_body(int);
for (auto i = 0; i < 100; ++i)
{
    q.parallel_for(range<1> {1024}, [&](item<1>& itm) { kernel_body(i); });
}
q.wait();
```
Then the user could instead submit 1 kernel that does N kernel body tasks concurrently:
``` C++
queue q;
void* kernel_body(int);
q.parallel_for(range<1> {1024}, [&](item<1>& itm)
{
    for (auto i = 0; i < 100; ++i) { kernel_body(i); }
});
q.wait();
```
[PRO] This transformation saves N-1 kernel submission overhead costs.

[CON] This transformation is fraught with exceptions -- it doesn't work with dependencies (an in-order queue, explicit events, USM deps, etc) -- it doesn't work with different sized/shaped ranges.

[CON] This transformation puts increased pressure on register space, memory working set, and so on.

> remove the acyclic restriction on a command_graph

You are right: that changes the execution model -- which makes it a different type of conversation :)